<a href="https://colab.research.google.com/github/Prajaktahz/Uni_Colab_Work/blob/main/FBA_Week_05_Python_Practice.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

![alt text](http://www.cs.nott.ac.uk/~pszgss/teaching/nlab.png)
# FBA Computing Session Week 5:

Python - Pandas Practice

Pandas is an incredibly useful library, and one that will likely be used by any business analyst who is working with Python. It is a set of functions that allow you to load in, cleanse, manipulate and descriptively analyse your data.

## Investigating University Admissions with and without Pandas

In our first couple of examples we are going to use admissions.csv (available on the Moodle). This file describes people who apply for a postgraduate degree at a US university, and indicates 3 relevant features about an applicant as well as whether they got into the Masters or PhD course they were applying for when they were selected by hand (note that if we were doing a full analytics project, this would be our target class!):

* The dataset has a binary output feature (i.e. dependent variable) called "admit".
* There are three predictor variables: gre, gpa and ranking.
* Variables gre (an exam result score) and gpa (the person’s grade point average) are continuous.
* The variable ranking takes on the values 1 through 4. Institutions with a rank of 1 have the highest prestige, while those with a rank of 4 have the lowest.

### Step A.1.
Before we had the <b>pandas</b> library we had to load data in using the built in "file" opening/closing commands (this remains important for you to have seen, even though we will generally use pandas in the future).

However, there is also a library called <b>"csv"</b> that's good know about, which simplifies parsing a csv with standard file commands. Let’s use that more traditional way first to load in the data we are using, and to understand who was admitted into a US university’s postgraduate program (and why), before we move on to pandas!

We’ll load it in from the admissions.csv file, using python’s csv library for now. Run the following code, but remember since we are working in Colab there are several ways to load files:

In [1]:
!wget -O week5_data.zip "https://drive.google.com/uc?export=download&id=18C-_-ojwxWq2HztrJaSBkQkrwvpMPszZ"
!unzip week5_data.zip

--2023-10-30 09:51:35--  https://drive.google.com/uc?export=download&id=18C-_-ojwxWq2HztrJaSBkQkrwvpMPszZ
Resolving drive.google.com (drive.google.com)... 173.194.216.113, 173.194.216.139, 173.194.216.102, ...
Connecting to drive.google.com (drive.google.com)|173.194.216.113|:443... connected.
HTTP request sent, awaiting response... 303 See Other
Location: https://doc-14-80-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/4651rck97si7aqa7nlf6r4inf4c691fv/1698659475000/02584936932483403665/*/18C-_-ojwxWq2HztrJaSBkQkrwvpMPszZ?e=download&uuid=d53ba4bf-ae68-4982-99fa-f1f65c6fb8ee [following]
--2023-10-30 09:51:35--  https://doc-14-80-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/4651rck97si7aqa7nlf6r4inf4c691fv/1698659475000/02584936932483403665/*/18C-_-ojwxWq2HztrJaSBkQkrwvpMPszZ?e=download&uuid=d53ba4bf-ae68-4982-99fa-f1f65c6fb8ee
Resolving doc-14-80-docs.googleusercontent.com (doc-14-80-docs.googleusercontent.com)... 142.251.162.132, 

In [2]:
import csv

#-- open the file with the data in it (UPDATE THIS!)
#-- loads the data in a "csv" object called a 'reader'
filename = "admissions.csv"

f = open(filename)

#-- loads the data in a "csv" object called a 'reader'
csv_data = csv.reader(f)

### Step A.2
You'll have seen above if the data loaded in correctly that in this method you can't just "print out" the data you load in, as it has landed in a data processing object (albeit a really useful object however, as it has recognized all the data's column names, and parsed all the commas, numbers and  words in the raw text held within the file into a <i>usable</i> format).

To get the data visibly out of that object, we just convert it to a list (which you should now be very familiar with):

In [3]:
#-- convert our data into a list to make it easier to examine:
data = list(csv_data)
print("Our data has", len(data), "lines")

Our data has 401 lines


Ok, so now the data is in a list, which you should be reasonably familiar with now, using a loop print out the first five lines of our data:

In [4]:
#-- Enter the rest of the code to print out the first 5 lines of the dataset:
for d in data[0:5]:
    print(d)

['admit', 'gre', 'gpa', 'ranking']
['0', '380', '3.61', '3']
['1', '660', '3.67', '3']
['1', '800', '4', '1']
['1', '640', '3.19', '4']


Ok you should see some datapoints!

But, if we're honest, there were a few inconveniences there. First, obviously we had to convert the data into a list. Second, you should have seen that the first line we printed out in the data is still the “header” with the feature names in it. Not ideal. And Third, you’ll notice that the data items are all still being viewed by python as text strings - when they are clearly numbers. Again, it will be a pain to do all the conversions manually (pandas will ultimately help us with this sort of thing). Nonetheless, let’s persevere with the csv library!

### Step A.3.
Let’s have a look at an individual field - specifically the grade point average column. Read through the following code which extracts data for a single column, converts it into a float, and then stores it in a new list:


In [8]:
#-- create a new list that is going to store just the gpa's columns data
gpa = []

#-- loop round every row apart from the first one (which was the header)
for d in data[1:]:

    #-- extract the element in the third (gpa) column  and convert to a float:
    value = float(d[2])

    #-- add the current value to our list of results:
    gpa.append(value)

#-- print out what we extracted:
print(gpa)

[3.61, 3.67, 4.0, 3.19, 2.93, 3.0, 2.98, 3.08, 3.39, 3.92, 4.0, 3.22, 4.0, 3.08, 4.0, 3.44, 3.87, 2.56, 3.75, 3.81, 3.17, 3.63, 2.82, 3.19, 3.35, 3.66, 3.61, 3.74, 3.22, 3.29, 3.78, 3.35, 3.4, 4.0, 3.14, 3.05, 3.25, 2.9, 3.13, 2.68, 2.42, 3.32, 3.15, 3.31, 2.94, 3.45, 3.46, 2.97, 2.48, 3.35, 3.86, 3.13, 3.37, 3.27, 3.34, 4.0, 3.19, 2.94, 3.65, 2.82, 3.18, 3.32, 3.67, 3.85, 4.0, 3.59, 3.62, 3.3, 3.69, 3.73, 4.0, 2.92, 3.39, 4.0, 3.45, 4.0, 3.36, 4.0, 3.12, 4.0, 2.9, 3.07, 2.71, 2.91, 3.6, 2.98, 3.32, 3.48, 3.28, 4.0, 3.83, 3.64, 3.9, 2.93, 3.44, 3.33, 3.52, 3.57, 2.88, 3.31, 3.15, 3.57, 3.33, 3.94, 3.95, 2.97, 3.56, 3.13, 2.93, 3.45, 3.08, 3.41, 3.0, 3.22, 3.84, 3.99, 3.45, 3.72, 3.7, 2.92, 3.74, 2.67, 2.85, 2.98, 3.88, 3.38, 3.54, 3.74, 3.19, 3.15, 3.17, 2.79, 3.4, 3.08, 2.95, 3.57, 3.33, 4.0, 3.4, 3.58, 3.93, 3.52, 3.94, 3.4, 3.4, 3.43, 3.4, 2.71, 2.91, 3.31, 3.74, 3.38, 3.94, 3.46, 3.69, 2.86, 2.52, 3.58, 3.49, 3.82, 3.13, 3.5, 3.56, 2.73, 3.3, 4.0, 3.24, 3.77, 4.0, 3.62, 3.51, 2.81,

Note, that the syntax <b>data[1:]</b> means use everything from line one onwards (i.e. not the zero-th line, which are the field names). If you ran the program successfully, then at least we can now see we have the gpa scores for each datapoint in their own list. Clearly not the simplest ways of getting at that data - however, it was good enough!

### Step A.4.
Finally, let’s do what we came for and find out some descriptive statistics about the grade point averages. Where’s good to start? Why mean, standard deviation, mins and maximums of course! The best library to use for any of this sort of thing is NumPy (which we will look at in more detail next week). In fact, as soon as anyone installs python they always add NumPy immediately. Read through and run the following program to see its mathematical functions:


In [9]:
import numpy as np
print("count", len(gpa))
print("mean", np.mean(gpa))
print("std", np.std(gpa))
print("min", np.min(gpa))
print("max", np.max(gpa))

count 400
mean 3.3899
std 0.3800907654758269
min 2.26
max 4.0


As a quick test enter the code below to create a different average, that is the optimal point predictor of a students GPA! (you may want to use google to search for the numpy documentation for it - or just take a stab yourself!)

In [10]:
#-- enter your code here...
print("median", np.median(gpa))

median 3.395


Well, ok good, but let’s be honest... that was all quite long-winded wasn’t it? Sure, you’re now able to use loops, load in files, create list variables and fill them (which is still a huge progress if you've not scripted before!) but it’s all somewhat of a hassle. Well let’s try and fix that by showing you how easy it can be in python too with the right analytics library! Introducing “Pandas”, friend of every real business analyst....

## Making things easier with Pandas
![pandas](https://drive.google.com/uc?export=view&id=1RxArfr4gjKs2u5QDIc7oAFIgVq4T32xF)

### Task B.1.
Ok, We’re going to do all of section A again. But this time … well, just type in the following (again adjusting your filename to the path where you have saved the admissions file):

In [11]:
import pandas
data = pandas.read_csv("admissions.csv")
print(data.head())
print(data.gpa.describe())

   admit  gre   gpa  ranking
0      0  380  3.61        3
1      1  660  3.67        3
2      1  800  4.00        1
3      1  640  3.19        4
4      0  520  2.93        4
count    400.000000
mean       3.389900
std        0.380567
min        2.260000
25%        3.130000
50%        3.395000
75%        3.670000
max        4.000000
Name: gpa, dtype: float64


Yup, that’s it.  All done. Just four lines to do everything we did in section A, and print it all to screen. Easier than excel even.

How have we done this? Well, Pandas is a python library that wraps up everything a budding analyst might need without jumping through all the hoops. As you should remember from the lecture, its <b>“read_csv()”</b> function creates a thing called a “dataframe” which just makes reading, accessing and getting summaries extremely easy.

Panda’s “head()” function shows you the top of your data so you can check to see it’s all looking okay… and notice it has automatically recognized what the field names are called from the “header” and removed it. ...and when we used <b>read_csv()</b> at the start it’s even converted the data from text into what it thinks the right data types are. Smooth.

And finally, we’ve got all those summary statistics by accessing the “gpa” column (magically python has converted that into the direct access of typing in <b>“data.gpa”</b>!) and just using the <b>“describe()”</b> method.

Ok, so given that’s so much easier why didn’t we just use this from the start!? Well:
* You need to know python basics - loops, converting data, etc. for when you ever hit problems.
* It’s good for the soul to learn the fundamentals. Full stop.
* Now you can see why analysts really use Python - people have added these fantastic libraries that make python scripting the quickest way to get solutions for real world business problems.

### Task B.2.

Let’s see all the statistics for the admission data. Change the describe() line so it examines every field:

In [12]:
# adjust that line of code to display a description of all columns
data.gpa.describe()

count    400.000000
mean       3.389900
std        0.380567
min        2.260000
25%        3.130000
50%        3.395000
75%        3.670000
max        4.000000
Name: gpa, dtype: float64

### Step B.3.
Pandas has clearly realised that our columns are not just text (otherwise it couldn’t have done the statistical analysis). It has tried to detect the correct data type for each column. It offers a very simple command to find out what it has picked:

In [13]:
print(data.dtypes)

admit        int64
gre          int64
gpa        float64
ranking      int64
dtype: object


If you see the column types it has detected (despite being only given raw text), you'll see it is not bad!

#### But which one has it got wrong? (don’t look ahead!)

### Step B.4
Assuming you’ve not just looked ahead (if you did, you cheated), you’ve now correctly realised that it’s the “admit” feature that is wrong. Because it has values of one and zero, even though they are always one or the other (a binary feature) pandas has decided that it’s a continuous numeric variable. Wrong of course - it’s a <b>category label</b>. They could have been just as easily “yes” and “no”, or “admitted” and “rejected”. There is nothing numerical about them. Pandas makes it very easy for us to tell python what the correct type should have been.

Run the following code:

In [14]:
data = pandas.read_csv(filename, dtype={'admit':'category'})

Check below that this has fixed the problem, by looking at the dtypes again:

In [16]:
#-- enter your code here
data.dtypes

admit      category
gre           int64
gpa         float64
ranking       int64
dtype: object

### Step B.5.
So far so good, but one final thing - we really want to show how each feature’s values differ for one class compared to the other (this would be useful for example, in a coursework who buy a product and don’t….. *cough*)

Let’s do this now by separating the positives (admitted - 1) from the negatives (rejected - 0) in our dataset, using the following the pandas syntax (just add it to the bottom of your program and run it again):


In [17]:
neg = data[data.admit == "0"]
pos = data[data.admit == "1"]
print(neg.describe())
print(pos.describe())

              gre         gpa     ranking
count  273.000000  273.000000  273.000000
mean   573.186813    3.343700    2.641026
std    115.830243    0.377133    0.917198
min    220.000000    2.260000    1.000000
25%    500.000000    3.080000    2.000000
50%    580.000000    3.340000    3.000000
75%    660.000000    3.610000    3.000000
max    800.000000    4.000000    4.000000
              gre         gpa     ranking
count  127.000000  127.000000  127.000000
mean   618.897638    3.489213    2.149606
std    108.884884    0.370177    0.917889
min    300.000000    2.420000    1.000000
25%    540.000000    3.220000    1.000000
50%    620.000000    3.540000    2.000000
75%    680.000000    3.755000    3.000000
max    800.000000    4.000000    4.000000


The first line creates a new slice of data with just the rows where the “admit” field was “0” representing people who were rejected (note that this must now be in quotes because it is a categorical label, and not a number), and the second line does the same for those rows representing people who were admitted into the university so the admit field is “1”. This way of filtering the data always reminds of SELECT in SQL (with the "where" clause inside the square brackets)... except obviously a lot quicker to type (SQL is so old…)

What this means is that we now can print summary statistics to compare between successful and unsuccessful applicants (the last two lines). Examine these.

What, if anything, can you see that is surprising or noteworthy to report back concerning these? (tip - check which feature is higher for admissions and rejections)


<b>What would you comment on concerning these comparisons between rejections and admissions to feedback to a client?:<b>

Total 273 students have faced rejection while 127 students have got accepted into the University.
Maximum marks in gre for students irrespective of getting admitted in university are 800, however the lowest scored by successful people is 300.

## Ice Cream Sales

Ok, now we have a new set of skills let's have a look again at our (now traditional) Ice Cream sales dataset, and use pandas to give us a descriptive analysis (n.b. we are repeating these processes, partly because these sorts of exploratory analysis are absolutely key for the beginning of any analytics reports... such as your courseworks).
Reload the dataset from Moodle and add to your google drive!

This time, use both the lecture slides, cheat sheets, and practice looking up documentation on google for the library. Good luck!

### Step C1. Import the Pandas library
Ok, to be certain we are not relying on previous jupyter cells, let's start by ensuring pandas is loaded in:

In [23]:
import pandas as pd
import numpy as np

### Step C2. Load the data

Load the data from the CSV file **ice_cream_sales.csv** using the Pandas function **read_csv()** and assign it to a variable **sales**.

Using the documentation for the method **read_csv()**, look for the parameter **index_col** to define the field 'year' as the index.

In [77]:
#-- enter your code here
sales = pd.read_csv('ice_cream_sales.csv', index_col='year')
sales.head()
#-- print out a check...
#print(sales)

january  february  march  april   may  june  july  august  september  \
year                                                                         
2020     1000      1100   1200   1300  1400  1500  1500    1400       1300   
2021     1100      1200   1300   1400  1500  1600  1600    1500       1400   
2022      500      1600   1700   1800  1900   500   500     500       1800   

      october  november  december  
year                               
2020     1200      1100      1000  
2021     1300      1200      1100  
2022     1700      1600       500

### Step C3. What are the total sales for each year?

Use the sum function, and check the documentation to specify the right axis (and ask for help if you hit a problem):

In [82]:
#-- enter your code here
# Need to explicitly use pandas here
print(sales.sum(axis=1))

'''for i in range(len(sales)):
  #print(i)
  df = sales.iloc[[i]]
  print(df.dtypes)'''

'''for index, row in sales.iterrows():
  print(index)
  print(row.sum())'''



year
2020    15000
2021    16200
2022    14600
dtype: int64


'for index, row in sales.iterrows():\n  print(index)\n  print(row.sum())'

### Step C4. What are the median sales for each year?

In [86]:
#-- enter your code here
sales.median(axis= 1) # '1' is for rows

year
2020    1250.0
2021    1350.0
2022    1600.0
dtype: float64

### Step C5. What is the average sale for each month across the 3 years?

In [85]:
#-- enter your code here
sales.mean(axis=0) # '0' is for columns

january       866.666667
february     1300.000000
march        1400.000000
april        1500.000000
may          1600.000000
june         1200.000000
july         1200.000000
august       1133.333333
september    1500.000000
october      1400.000000
november     1300.000000
december      866.666667
dtype: float64

## A survey of users and their occupation

And for a final bit of practice let's try another data file, this time based on people's occupations...

### Step D1. Import the necessary libraries

In [ ]:
#-- Pandas should already be imported from the previous sections.
#-- If you haven't you will need to import it here


### Step D2. Assign the data to a variable called users and use the 'user_id' as index

The data to use for this exercise is the CSV file 'occupation.csv'. Using the Pandas function read_csv(), load the data in a variable called users and use the 'user_id' as Index.

***A good reflex to have when you work with a new dataset is to have a quick look at the data files you were given to see how they are structured. You don't have to use Python for that, simply open it in a text editor, Excel, ...***

In [87]:
users = pd.read_csv('occupation.csv', sep=';', index_col='user_id')

### Step D3. See the first 25 entries

In [88]:
#-- enter your code here...
users.head(25)

age gender     occupation zip_code
user_id                                    
1         24      M     technician    85711
2         53      F          other    94043
3         23      M         writer    32067
4         24      M     technician    43537
5         33      F          other    15213
6         42      M      executive    98101
7         57      M  administrator    91344
8         36      M  administrator    05201
9         29      M        student    01002
10        53      M         lawyer    90703
11        39      F          other    30329
12        28      F          other    06405
13        47      M       educator    29206
14        45      M      scientist    55106
15        49      F       educator    97301
16        21      M  entertainment    10309
17        30      M     programmer    06355
18        35      F          other    37212
19        40      M      librarian    02138
20        42      F      homemaker    95660
21        26      M         writer    30068
22        25      M         writer    40206
23        30      F         artist    48197
24        21      F         artist    94533
25        39      M       engineer    55107

### Step D4. See the last 10 entries

In [89]:
#-- enter your code here...
users.tail(10)

age gender     occupation zip_code
user_id                                    
934       61      M       engineer    22902
935       42      M         doctor    66221
936       24      M          other    32789
937       48      M       educator    98072
938       38      F     technician    55038
939       26      F        student    33319
940       32      M  administrator    02215
941       20      M        student    97229
942       48      F      librarian    78209
943       22      M        student    77841

### Step D5. What is the number of observations in the dataset?

Note that there are a couple of ways of doing this (you might, for example leverage the keywords shape or len!)

In [92]:
#-- enter your code here
print(users.shape)
print(len(users))

(943, 4)
943


### Step D6. What is the number of columns in the dataset?

In [99]:
#-- enter your code here
#users.columns
len(users.axes[1])

4

### Step D7. Print the name of all the columns.

In [102]:
#-- enter your code here
users.columns

Index(['age', 'gender', 'occupation', 'zip_code'], dtype='object')

### Step D8. How is the dataset indexed?

A Pandas dataframe always has an Index. If you do not specify a specific one, Pandas will automatically create one which is going to be the row number.

In [ ]:
#-- "the index" (aka "the labels")

### Step D9. What is the data type of each column?

As before, using the **.dtypes** attribute of your **users** dataframe, you can have a list of all your columns and their data type.

In [ ]:
#-- enter your code here
print(users.dtypes)

### Step D10. Print only the occupation column

In [ ]:
#-- enter your code here

### Step D11. How many different occupations there are in this dataset?

In [ ]:
#-- enter your code here

### Step D12. What is the most frequent occupation?

In [ ]:
#-- enter your code here

### Step D13. Summarise the DataFrame.

Remember, you can summarise the values of a DataFrame using the **.describe()** method. Do so now...

In [ ]:
#-- enter your code here

### Step D14. Summarise all the columns

However, the **.describe()** method only considers numeric columns by default. If you want to include the columns with another data type, you can pass the argument **include = "all"** to the method.

In [ ]:
#-- enter your code here

### Step D15. Summarise only the occupation column

In [ ]:
#-- enter your code here

### Step D16. What is the mean age of users? and the median?

In [ ]:
#-- enter your code here, rounding to the nearest whole number...

### Step D17. What is the age with least occurrence?

Try to use the "value counts()" and "tail()" methods to solve this problem...

In [ ]:
#-- enter your code here

### Step D18. What are the main occupations of users under 20 years old?

Use value_counts() again, but this time filter your data for users who are under the age of twenty. There are many ways to do this but a clever bit of pandas indexing is probably the neatest.

In [ ]:
#-- enter your code here

Congratulations, you are well on your way to being proficient in using Pandas for your business analytics problems (as we will all year in fact). And don't fret if you forget the syntax - this is normal. The skill is being able to look it up again quickly!